src/Language/Pirouette/QuasiQuoter/Internal.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DeriveLift #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Language.Pirouette.QuasiQuoter.Internal where
+
+import Control.Monad.Except (runExcept)
+import Control.Monad.Reader
+import qualified Data.Map as M
+import Language.Haskell.TH.Quote
+import Language.Haskell.TH.Syntax hiding (Name, Type)
+import Language.Pirouette.QuasiQuoter.Syntax
+import Language.Pirouette.QuasiQuoter.ToTerm
+import Pirouette.Term.Syntax.Base
+import Pirouette.Term.Syntax.Pretty.Class (Pretty (..))
+import qualified Pirouette.Term.Syntax.SystemF as SystF
+import Text.Megaparsec
+
+parseQ :: Parser a -> String -> Q a
+parseQ p str =
+  case runReader (runParserT p "<template-haskell>" str) Nothing of
+    Left err -> fail (errorBundlePretty err)
+    Right r -> return r
+
+trQ :: TrM a -> Q a
+trQ f = case runExcept f of
+  Left err -> fail err
+  Right r -> return r
+
+maybeQ :: (Pretty e) => Either e a -> Q a
+maybeQ (Left e) = fail $ show $ pretty e
+maybeQ (Right x) = return x
+
+instance (Ord k, Lift k, Lift v) => Lift (M.Map k v) where
+  liftTyped m = [||M.fromList $$(liftTyped (M.toList m))||]
+
+quoter :: (String -> Q Exp) -> QuasiQuoter
+quoter quote =
+  QuasiQuoter
+    { quoteExp = quote,
+      quotePat = \_ -> failure0 "patterns",
+      quoteType = \_ -> failure0 "types",
+      quoteDec = \_ -> failure0 "declarations"
+    }
+  where
+    failure0 k =
+      fail ("this quasiquoter does not support splicing " ++ k)
+
+-- Orphan instances to 'Lift' the necessary types
```

Comment:
Would `{-# OPTIONS_GHC -Wno-orphans #-}` result in cleaner builds (eventually, even if this warning is disabled project-wide now)?

Reply:
Frankly, I am not a good enough Haskeller to have any idea whether it would improve the situation or not. I added it in https://github.com/tweag/pirouette/pull/168/commits/34d8e7930b43009ec299916ccc20f6ee57fec134.

Reply:
This does raise another question for me: why did we disable the warning on a project level if it's easy to disable it only on some specific well-defined files?

Reply:
> Frankly, I am not a good enough Haskeller to have any idea whether it would improve the situation or not.

Ideally, we'd define these instances in the module defining the types in question, especially if we expect them to be used in TH contexts.

> This does raise another question for me: why did we disable the warning on a project level if it's easy to disable it only on some specific well-defined files?

Dunno, that's a good question! I personally tend to disable only those warnings that I expect to occur in more or less every module in the project.